[dogstatsd] enable the no-aggregation pipeline.

remeh opened the pull request:
### What does this PR do?

Enable the DogStatsD no-aggregation pipeline by default.

### Motivation

Remove friction for users to enable this feature.

### Additional notes

Some unit tests in `corechecks` are modifying the internal buffer of the shared logger or some configuration stuff. While doing so, the unit test may also be initing the demultliplexer/aggregator causing races. In order to avoid this, all tests using the mocked Sender (used for checks sample collection, not related to the no-agg pipeline in any way) are not enabling the no-aggregation pipeline.

### Possible Drawbacks / Trade-offs

This change is expected to have no impact: a new goroutine will run the no-aggregation pipeline but benchmarks have measured that it has no impact on the rest of the Agent (RSS, CPU, ...) when not receiving traffic.

This said, I'm adding the `major_change` label since it is transversal to all Agents using an `AgentDemultiplexer` implementation.

### Describe how to test/QA your changes

* Make sure the Datadog Agent is still processing DogStatsD traffic.
* Send a DogStatsD packet 2h late `echo -n "custom_metric:60|g|#env:test,dev:remy|T$(( $(date +%s) - 7200 ))" | nc -4u -w0 127.0.0.1 8125`
* Validate your metric appear in the Datadog app
* Generate a flare of the running Agent and validate that the file `expvar/no_aggregation` contains sensible information.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

remeh committed:
[dogstatsd] enable the no-aggregation pipeline.

remeh committed:
Merge remote-tracking branch 'origin/main' into remeh/enable-no-agg

remeh committed:
[dogstatsd] adapt the unit-test testing the no-aggregation pipeline default state in the demux.

remeh committed:
[demux] trying to remove a race happening in the seelog logger at startup

cswatt commented on docs/dogstatsd/internals.md:
```diff
@@ -92,3 +92,12 @@ The following calculations determine the number of TimeSamplerWorker and TimeSam
+
+Input: slice of MetricSamples
+
+The NoAggregationStreamWorker runs an infinite loop in a goroutine, it receives metric sample with timestamp and batches them to send them as fast as possible to the intake. It performs no aggregation nor extra processing except from adding tags to the metrics.
```
```suggestion
The NoAggregationStreamWorker runs an infinite loop in a goroutine. It receives metric samples with timestamps, and it batches them to be sent as quickly as possible to the intake. It performs no aggregation nor extra processing, except from adding tags to the metrics.
```

cswatt commented on docs/dogstatsd/internals.md:
```diff
@@ -92,3 +92,12 @@ The following calculations determine the number of TimeSamplerWorker and TimeSam
+
+It is running only when `dogstatsd_no_aggregation_pipeline` is set to `true`.
+
+Payload it is sending to the intake (through the normal `Serializer`/`Forwarder` pieces) contains at maximum `dogstatsd_no_aggregation_pipeline_batch_size` metrics, default is `256`.
```
```suggestion
The payload being sent to the intake (through the normal `Serializer`/`Forwarder` pieces) contains, at maximum, `dogstatsd_no_aggregation_pipeline_batch_size` metrics. This value defaults to `256`.
```

cswatt reviewed (commented):
docs editing

remeh committed:
Apply suggestions from code review

Co-authored-by: cecilia saixue watt <cecilia.watt@datadoghq.com>

remeh committed:
[collector] fix races in snmp unit tests while using seelog.

They are modifying the logger on the fly during the unit tests, which is not to
the like of other parts running concurrently and using the logger.

remeh committed:
Merge branch 'remeh/enable-no-agg' of github.com:DataDog/datadog-agent into remeh/enable-no-agg

remeh committed:
[collector] do not run the no-agg worker while snmp tests are running

The snmp tests are setuping a different logger, causing race with a no-aggregation
worker which is logging.

remeh committed:
[collector] all tests using mock senders (i.e. checks) does not enable the no agg pipeline.

Some of them are modifying the internal buffer of the shared logger or some configuration stuff. While doing so, the unit test may also be initing the demultliplexer/aggregator causing races. In order to avoid this, all tests using the mock Sender (used with Checks collector) are not enabling the no-aggregation pipeline.
